pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -756,3 +762,30 @@ func (manager *gceServiceManager) getProjectsAPIEndpointAlpha() string {
 
 	return projectsApiEndpoint
 }
+
+// GetAllCurrentZones returns all the zones in which nodes are currently running
+// TODO: Caching, but this is currently only called when we are creating a volume,
+// which is a relatively infrequent operation, and this is only 1 API call.
+// Ideally we want to make this watch-based
+// TODO: When k8s node labels are secure we should use those.
+func (manager *gceServiceManager) GetAllCurrentZones(client clientset.Interface) (sets.String, error) {
+	zones := sets.NewString()
+	if client == nil {
+		return nil, fmt.Errorf("Client not set on GCECloud object")
+	}
+	nodeList, err := client.Core().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, node := range nodeList.Items {
+		labels := node.ObjectMeta.Labels
+		zone, ok := labels[kubeletapis.LabelZoneFailureDomain]
```

Comment:
Why not just make the list call above use a labelSelector that checks that the zone label exists?

Reply:
Great suggestion!

Reply:
The only downside is that instead of one api call, we need to make one for each zone in the region every time.  I don't think create disk operations are in practice frequent enough to hit qps limits (although it's easy to come up with a theoretical example that will), but it's something to be aware of.  If we make this informer based, then this problem should be eliminated.

Reply:
I discussed with @davidz627, and we decided to look into the informer approach, and keep a cache of zones in the cloud provider that gets updated on node add/delete events.  This solves all 3 issues:

* doesn't make any additional API calls
* don't need to iterate through all the nodes
* don't need to add node list permissions for cloud provider??? (because controller manager already has list permissions for the informer)  Actually I'm not sure why we needed to add the cluster role binding in the first place as controller manager should already have list permissions.

Reply:
Perfect!
Regarding your 3rd point, I guess it's because each controller has its own identity and hence its own RBAC (https://github.com/davidz627/kubernetes/blob/a60e757f816d7468e805fa9500b83c344576c3b6/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml).

Reply:
One possible issue.  Going with the informer method requires us to modify the cloud provider interface to be able to pass in the informer.  But I think the cloud provider API is under a freeze as we're migrating to out of tree cloud providers.